Recipes/Protocols.md

```diff
@@ -53,3 +53,80 @@ class MyClass: MyProtocol {
     }
 }
 ```
+
+### Solutions #3: Create a wrapper around the protocol
+If the protocol you want to confirm to is not under your control, and the protocol is non-isolated, you can create a non-concurrent wrapper around the protocol and then create an async protocol for your actor. For example, URLSession delegates that are not async.
+E.g:
+
+```swift
+// This could also be an actor instead of a class.
+@MainActor
+class MyClass {
+    var myClassProperty = "pew"
+    let someObject: Object
+
+    init() {
+        someObject = Object()
+        someObject.delegate = self // <----- We want to use the object delegate calls
+    }
+}
+
+extension MyClass: ObjectNonConcurrentDelegate { // <------- Not the best way ❗
+    // ERROR!: You would need to mark the method nonisolated like Solution #1.
+    // But sometimes we want to access our class properites, which means we would
+    // be accessing our actor from outside the actor system/boundries and this could lead to
+    // data races and threading issues, which is what we want to avoid
+    // when we use actors in the first place!
```

Comment:
I think this comment is misleading. Given MyClass is marked @MainActor, Solution #1 would actually be simpler, correct, and have no restrictions.

However! That adds the *huge* restriction that all delegate calls by `someObject` actually happen on the MainActor. This may or may not be the case. I have run into many Objective-C classes that do not mark their delegate protocols as Sendable but actually, in practice, do call their delegate methods on other threads.

What do you think about changing this example from a class to an actor to really demonstrate that you want to support that non-MainActor behavior>

Reply:
ummm, good point. The thing is, I don't think there is now a solution that could give you half of Solution #1 and half of my solution. So if neither of the protocol/delegate method access the actor's properties, then solution #1 is the go. But if they do access the actor, then my solution works. But I guess the case where some of methods do and some don't is tricky.

I will change my example to be an `actor` as you said. What do you think the solution could be in the case I mentioned?

Also, I feel like my solution and solution #1 work best if you don't have control over the protocol's methods and you have to choose to mark them as `nonisolated` and stay away from the actor's properties, or find another way.

Reply:
This is a hard problem.

An actor can become a delegate *only* if the delegate methods never return anything *and* all arguments are Sendable. I don't think it is possible to do it safely in any other way.

If the delegate is MainActor-only, then solution #1 is the easiest option, and why are you using an actor for a MainActor-only task anyways?

However, there is a third possibility - the delegate is *not* MainActor-only. This is a situation you have to handle extremely carefully, and may be impossible to support safely without additional work from whoever controls the delegate protocol.

Reply:
> An actor can become a delegate only if the delegate methods never return anything and all arguments are Sendable. I don't think it is possible to do it safely in any other way.

I completely agree, at least for the time being. The whole actor system only works as intended if utilized the way it should, by using Sendable and carefully following the actor system.

> However, there is a third possibility - the delegate is not MainActor-only. This is a situation you have to handle extremely carefully, and may be impossible to support safely without additional work from whoever controls the delegate protocol.

I haven't thought about this one. Can you please give me an example of this? I will have to think about it.
Also, do you think we should include your comments and observations in my solution?

Reply:
> Can you please give me an example of this?

`NSURLSession` is a great example! The isolation of its delegate is dependent on how it is initialized. In short, its delegate is fundamentally incompatible with Swift concurrency in the general case.

https://developer.apple.com/documentation/foundation/nsurlsession/1411597-sessionwithconfiguration

This pattern comes up over and over again. Objective-C is not subject to any of the Swift concurrency rules, so it is trivial to make APIs that *cannot* be used safely from Swift 😅

Reply:
I mean, it is after all called the "Swift Concurrency" system so 🤷‍♂️
Regardless, I feel believe the solution I provided can be beneficial to solve some of the issues one might face when using actors. If you don't have any more comments on it or changes, I think we can proceed to merge and it to the recipes.

I think we can also open an issue for discussion with the third possibility example you provided.